### DISCUSSION TITLE
Symfony serializer (v6.1.6) slow denormalize operation

### DISCUSSION THREAD

**Discussion by itodor** [symfony/symfony#48863] [Q&A]
I am having a simple REST API developed with Symfony 6. I am using the following serializer for my application:

```$encoders = [new JsonEncoder()];
$extractor = new PropertyInfoExtractor([], [new PhpDocExtractor(), new ReflectionExtractor()]);
$responseNormalizer = new ObjectNormalizer(null, new CamelCaseToSnakeCaseNameConverter(),null, $extractor);
$dateTimeNormalizer = new DateTimeNormalizer([DateTimeNormalizer::FORMAT_KEY => DateTimeType::DEFAULT_FORMAT]);
$arrayDenormalizer = new ArrayDenormalizer();
$normalizers = [$arrayDenormalizer, $dateTimeNormalizer, $responseNormalizer];
$serializer = new Serializer($normalizers, $encoders);

return $serializer->denormalize($data, $className);
```

I realized my response times are very slow and figured out that the ->denormalize() call is the bottleneck. Even though these tests were made on local environment I noticed that when I make simple GET request for products, this returns 20 items per page (paginated response) and denormalize call for each of these 20 items takes around 300-400ms on average.

My question is, is there anything I am doing obviously wrong, a way to optimize my Serializer definition? I found in couple of posts that ObjectNormalizer is the problematic slow one but didn't manage to find any concrete solutions how to optimize it?

**Comment by WebMamba** [comment]
Do you get the object you denormalize from doctrine? Denormalize can be slow if you have a big entity with a lot of relations. If you want to have a deeper look into the denormalize process, you can use the web debug toolbar 😁
